CSCDESTINCLIMADT-1150: sqlite3.OperationalError: too many SQL variables

### Description

From @youcheng-csc in DestinE JIRA CSCDESTINCLIMADT-1150.

The command `autosubmit setstatus o027 -fs WAITING -t COMPLETED -s -np` for that experiment resulted in this:

```bash
[ERROR] 2026-03-18 18:47:20,309 In Configuration files:
Issues in [Jobs] config file:
[INI] Dependency parameter is invalid, job COMPILE is not configured
[eCode=6013] 
```

and

```bash
sqlite3.OperationalError: too many SQL variables 
```

* Autosubmit version: 4.1.16
* Machine/VM/environment name (if applicable): destine
* Experiment ID (if applicable): o027
* Experiment tasks & log path if applicable): /appl/AS/AUTOSUBMIT_DATA/o027/tmp/ASLOGS/20260318_184715_setstatus_err.log

### Reproducible Example

<!--
  Please provide an example of how to replicate this bug to help us diagnose
  the issue.
-->
TODO

### Expected Behaviour

<!-- Please describe what you expected to happen. -->
Either set status correctly, or a useful message.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -15,6 +15,7 @@
 - Fixed exception trace in AutosubmitCritical and AutosubmitError not being displayed to the user #2109
 - Fix `Jobs.running` to case-insensitive #2916
 - Fixed the wallclock validation error message to display the platform's wallclock time instead of the job time twice #2492
+- Fixed recovery and setstatus for experiments with more than > 32k jobs #2894
 
 **Enhancements:**
 

--- a/autosubmit/history/database_managers/experiment_history_db_manager.py
+++ b/autosubmit/history/database_managers/experiment_history_db_manager.py
@@ -17,6 +17,7 @@
 
 import os
 import textwrap
+import time
 from contextlib import contextmanager
 from pathlib import Path
 from typing import Any, Generator, Optional, Protocol, cast
@@ -889,14 +890,15 @@ class SqlAlchemyExperimentHistoryDbManager:
         # Local import avoids shadowing any `table` / `column` parameter.
         from sqlalchemy import column as _col, table as _tbl
 
-        tmp = _tbl("_tmp_job_names", _col("job_name"))
+        table_name = f"_tmp_job_names_{time.time_ns()}"
+        tmp = _tbl(table_name, _col("job_name"))
         with self.engine.connect() as conn:
             conn.execute(text(
-                "CREATE TEMPORARY TABLE IF NOT EXISTS _tmp_job_names (job_name TEXT)"
+                f"CREATE TEMPORARY TABLE IF NOT EXISTS {table_name} (job_name TEXT)"
             ))
-            conn.execute(text("DELETE FROM _tmp_job_names"))
+            conn.execute(text(f"DELETE FROM {table_name}"))
             conn.execute(
-                text("INSERT INTO _tmp_job_names (job_name) VALUES (:name)"),
+                text(f"INSERT INTO {table_name} (job_name) VALUES (:name)"),
                 [{"name": n} for n in job_names],
             )
             try:
@@ -905,7 +907,7 @@ class SqlAlchemyExperimentHistoryDbManager:
                 conn.rollback()
                 raise
             finally:
-                conn.execute(text("DROP TABLE IF EXISTS _tmp_job_names"))
+                conn.execute(text(f"DROP TABLE IF EXISTS {table_name}"))
                 conn.commit()
 
     def select_jobs_data(self, table, job_names: list[str]) -> list[tuple[str, Any]]:

--- a/autosubmit/history/database_managers/experiment_history_db_manager.py
+++ b/autosubmit/history/database_managers/experiment_history_db_manager.py
@@ -17,10 +17,11 @@
 
 import os
 import textwrap
+from contextlib import contextmanager
 from pathlib import Path
-from typing import Any, Optional, Protocol, cast
+from typing import Any, Generator, Optional, Protocol, cast
 
-from sqlalchemy import and_, func, inspect, desc, insert, select, update
+from sqlalchemy import and_, func, inspect, desc, insert, select, text, update
 from sqlalchemy.schema import CreateTable, CreateSchema
 
 import autosubmit.history.utils as HUtils
@@ -331,10 +332,10 @@ class ExperimentHistoryDbManager(DatabaseManager):
                 failed, queuing, running, 
                 submitted, suspended, metadata) VALUES(?,?,?,?,?,?,?,?,?,?,?,?,?,?) '''
         arguments = (
-        HUtils.get_current_datetime(), HUtils.get_current_datetime(), experiment_run.start, experiment_run.finish,
-        experiment_run.chunk_unit, experiment_run.chunk_size, experiment_run.completed, experiment_run.total,
-        experiment_run.failed, experiment_run.queuing, experiment_run.running,
-        experiment_run.submitted, experiment_run.suspended, experiment_run.metadata)
+            HUtils.get_current_datetime(), HUtils.get_current_datetime(), experiment_run.start, experiment_run.finish,
+            experiment_run.chunk_unit, experiment_run.chunk_size, experiment_run.completed, experiment_run.total,
+            experiment_run.failed, experiment_run.queuing, experiment_run.running,
+            experiment_run.submitted, experiment_run.suspended, experiment_run.metadata)
         return self.insert_statement_with_arguments(self.historicaldb_file_path, statement, arguments)
 
     def update_many_job_data_change_status(self, changes):
@@ -364,7 +365,7 @@ class ExperimentHistoryDbManager(DatabaseManager):
             job_data_dc.job_id, job_data_dc.status, job_data_dc.energy, job_data_dc.extra_data,
             job_data_dc.nnodes, job_data_dc.ncpus, job_data_dc.rowstatus, job_data_dc.out, job_data_dc.err,
             job_data_dc.children, job_data_dc.platform_output, job_data_dc._id, job_data_dc.workflow_commit, job_data_dc._id
-            )
+        )
         self.execute_statement_with_arguments_on_dbfile(self.historicaldb_file_path, statement, arguments)
 
     def _update_experiment_run(self, experiment_run_dc):
@@ -871,15 +872,52 @@ class SqlAlchemyExperimentHistoryDbManager:
                 jobs_data_by_name[job['job_name']] = job
         return jobs_data_by_name
 
-    def select_jobs_data(self, table, job_names) -> list[tuple[str, Any]]:
-        query = select(table).where(
-            and_(
-                table.c.last == 1,
-                table.c.job_name.in_(job_names)
-            )
-        )
+    @contextmanager
+    def _job_names_tmp_table(
+            self, job_names: list[str]
+    ) -> Generator[tuple[Any, Any], None, None]:
+        """Context manager that populates a temporary table with job names.
+
+        Yields ``(conn, tmp)`` where ``conn`` is the active SQLAlchemy connection
+        and ``tmp`` is a table expression for ``_tmp_job_names``.  The table is
+        created, cleared, populated, and dropped automatically, so callers only
+        need to write the query itself.
+
+        :param job_names: Job names to load into the temporary table.
+        :yields: ``(conn, tmp)`` — the connection and the temp table expression.
+        """
+        # Local import avoids shadowing any `table` / `column` parameter.
+        from sqlalchemy import column as _col, table as _tbl
+
+        tmp = _tbl("_tmp_job_names", _col("job_name"))
         with self.engine.connect() as conn:
-            rows = conn.execute(query).fetchall()
+            conn.execute(text(
+                "CREATE TEMPORARY TABLE IF NOT EXISTS _tmp_job_names (job_name TEXT)"
+            ))
+            conn.execute(text("DELETE FROM _tmp_job_names"))
+            conn.execute(
+                text("INSERT INTO _tmp_job_names (job_name) VALUES (:name)"),
+                [{"name": n} for n in job_names],
+            )
+            try:
+                yield conn, tmp
+            finally:
+                conn.execute(text("DROP TABLE IF EXISTS _tmp_job_names"))
+                conn.commit()
+
+    def select_jobs_data(self, table, job_names: list[str]) -> list[tuple[str, Any]]:
+        """Return last=1 job_data rows for the requested job names.
+
+        :param table: The SQLAlchemy ``job_data`` Table object to query.
+        :param job_names: Job names to look up.
+        :return: List of ``(column_name, value)`` tuples, one per matching row.
+        """
+        with self._job_names_tmp_table(job_names) as (conn, tmp):
+            rows = conn.execute(
+                select(table)
+                .join(tmp, table.c.job_name == tmp.c.job_name)
+                .where(table.c.last == 1)
+            ).fetchall()
         columns = table.c.keys()
         return [tuple(zip(columns, row)) for row in rows]
 

--- a/autosubmit/history/database_managers/experiment_history_db_manager.py
+++ b/autosubmit/history/database_managers/experiment_history_db_manager.py
@@ -901,6 +901,9 @@ class SqlAlchemyExperimentHistoryDbManager:
             )
             try:
                 yield conn, tmp
+            except Exception:
+                conn.rollback()
+                raise
             finally:
                 conn.execute(text("DROP TABLE IF EXISTS _tmp_job_names"))
                 conn.commit()

--- a/test/unit/history/database_managers/test_experiment_history_db_manager.py
+++ b/test/unit/history/database_managers/test_experiment_history_db_manager.py
@@ -15,11 +15,17 @@
 # You should have received a copy of the GNU General Public License
 # along with Autosubmit.  If not, see <http://www.gnu.org/licenses/>.
 
+import sqlite3
+
 import pytest
+from sqlalchemy import and_, insert, select
 
 from autosubmit.config.basicconfig import BasicConfig
+from autosubmit.database.tables import JobDataTable, get_table_with_schema
+from autosubmit.history.utils import get_current_datetime
 from autosubmit.history.database_managers.experiment_history_db_manager import (
-    create_experiment_history_db_manager, SqlAlchemyExperimentHistoryDbManager
+    create_experiment_history_db_manager,
+    SqlAlchemyExperimentHistoryDbManager,
 )
 
 
@@ -41,3 +47,62 @@ def test_functions_not_implemented(mocker):
     ]:
         with pytest.raises(NotImplementedError):
             getattr(db_manager, fn)()
+
+
+def test_select_jobs_data_regression_sqlite_variable_limit(tmp_path, monkeypatch):
+    """Regression test: old IN-clause crashes above SQLite's variable-number limit.
+        The default limit varies by Python version and SQlite build, it is often 999 or 32766
+    """
+    monkeypatch.setattr(BasicConfig, 'DATABASE_BACKEND', 'sqlite')
+
+    with sqlite3.connect(":memory:") as raw:
+        # Python 3.11+ exposes getlimit(); older builds default to 999 or 32766.
+        sqlite_limit = raw.getlimit(9) if hasattr(raw, 'getlimit') else 32766
+
+    n_jobs = sqlite_limit + 1
+
+    db_manager = SqlAlchemyExperimentHistoryDbManager(
+        schema="test_limit",
+        jobdata_path=str(tmp_path),
+        jobdata_file="job_data_test_limit.db",
+    )
+    db_manager.initialize()
+
+    job_names = [f"test_limit_20200101_fc0_{i}_SIM" for i in range(1, n_jobs + 1)]
+    job_data_table = get_table_with_schema(db_manager.schema, JobDataTable)
+    now = get_current_datetime()
+
+    rows = [
+        {
+            "counter": 1, "job_name": name, "created": now, "modified": now,
+            "submit": 0, "start": 0, "finish": 0, "status": "COMPLETED",
+            "rowtype": 0, "ncpus": 0, "wallclock": "00:00", "qos": "debug",
+            "energy": 0, "date": "20200101", "section": "SIM", "member": "fc0",
+            "chunk": idx + 1, "last": 1, "platform": "LOCAL", "job_id": idx + 1,
+            "extra_data": "{}", "nnodes": 0, "run_id": 1,
+            "MaxRSS": 0.0, "AveRSS": 0.0, "out": "", "err": "",
+            "rowstatus": 0, "children": None, "platform_output": None,
+        }
+        for idx, name in enumerate(job_names)
+    ]
+
+    with db_manager.engine.connect() as conn:
+        conn.execute(insert(job_data_table), rows)
+        conn.commit()
+
+    # Old approach: IN-clause with N bound parameters must fail above the limit.
+    old_query = select(job_data_table).where(
+        and_(
+            job_data_table.c.last == 1,
+            job_data_table.c.job_name.in_(job_names),
+        )
+    )
+    with db_manager.engine.connect() as conn:
+        with pytest.raises(Exception, match="too many SQL variables|SQLITE_MAX_VARIABLE_NUMBER"):
+            conn.execute(old_query).fetchall()
+
+    # New approach
+    result = db_manager.select_jobs_data(job_data_table, job_names)
+
+    assert len(result) == n_jobs
+    assert all(dict(row)["last"] == 1 for row in result)

--- a/test/unit/history/database_managers/test_experiment_history_db_manager.py
+++ b/test/unit/history/database_managers/test_experiment_history_db_manager.py
@@ -56,9 +56,8 @@ def test_select_jobs_data_regression_sqlite_variable_limit(tmp_path, monkeypatch
     monkeypatch.setattr(BasicConfig, 'DATABASE_BACKEND', 'sqlite')
 
     with sqlite3.connect(":memory:") as raw:
-        # Python 3.11+ exposes getlimit(); older builds default to 999 or 32766.
-        sqlite_limit = raw.getlimit(9) if hasattr(raw, 'getlimit') else 32766
-
+        # Python 3.11+ exposes getlimit() default seems to be 250000; older builds default to unknown number (32766 fails in the CI/CD)
+        sqlite_limit = raw.getlimit(9) if hasattr(raw, 'getlimit') else 250000
     n_jobs = sqlite_limit + 1
 
     db_manager = SqlAlchemyExperimentHistoryDbManager(

